apple/internal/xcarchive/make_xcarchive.py

```diff
@@ -0,0 +1,116 @@
+    )
+    parser.add_argument(
+        "--dsym",
+        required=False,
+        action="append",
+        help="The file path to a dSYM file which will be packaged in the .xcarchive.",
+    )
+    return parser
+
+def _main(
+        infoplist_path: str,
+        archive_path: str,
+        output_path: str,
+        dsyms: list):
+    # Construct the .xcarchive.
+    # <BundleName>.xcarchive
+    #   Products/
+    #     Applications/
+    #       <BundleName>.app
+    #   Info.plist
+    #   PkgInfo
+    #   dSYMs/
+    #     <BundleName>.app.dSYM
+    #
+
+    # Find the .app path, if the archive is an .ipa this will be within the Payload directory.
+    # We'll need to unzip the .ipa to find the .app, if we're given an .app this will be a no-op.
+    if archive_path.endswith(".ipa"):
+        with open(archive_path, "rb") as f:
+            archive = zipfile.ZipFile(f)
+            archive.extractall()
+            app_path = next(Path().glob("**/*.app"))
+            if not app_path:
+                raise Exception("Could not find .app within .ipa located at: " + archive_path)
+            archive_path = str(app_path)
+
+    # Collect the required information for the .xcarchive plist
+    # from the apps Info.plist.
+    with open(infoplist_path, "rb") as f:
+        infoplist = plistlib.load(f, fmt=plistlib.FMT_BINARY)
+        bundle_name = infoplist.get("CFBundleDisplayName", infoplist["CFBundleName"])
+        bundle_id = infoplist["CFBundleIdentifier"]
+        bundle_version = infoplist["CFBundleVersion"]
+        short_version = infoplist["CFBundleShortVersionString"]
+
+    # Create the .xcarchive directory.
+    os.makedirs(output_path, exist_ok=True)
+
+    # Copy the app bundle into the .xcarchive.
+    destination = os.path.join(
+        output_path,
+        "Products",
+        "Applications",
+        os.path.basename(archive_path))
+    os.makedirs(destination, exist_ok=True)
+    shutil.copytree(archive_path, destination, dirs_exist_ok=True)
+
+    # Copy the dSYM files into the .xcarchive.
+    for dsym in dsyms or []:
+        shutil.copytree(
+            dsym,
+            os.path.join(output_path, "dSYMs", os.path.basename(dsym)),
+            dirs_exist_ok=True)
+
+    # Create the Info.plist for the .xcarchive.
+    info_plist = {
+        "ApplicationProperties": {
+            "ApplicationPath": os.path.join(
+                "Products",
+                "Applications",
+                os.path.basename(archive_path)),
+            "CFBundleIdentifier": bundle_id,
+            "CFBundleShortVersionString": short_version,
+            "CFBundleVersion": bundle_version,
+        },
+        "Name": bundle_name,
+        "SchemeName": bundle_name,
+        "CreationDate": datetime.datetime.now().isoformat(),
+    }
+    info_plist_path = os.path.join(output_path, "Info.plist")
+    with open(info_plist_path, "wb") as f:
+        plistlib.dump(info_plist, f)
```

Comment:
Is that just creating the directory before-hand? I added that to the start of `_main`

Reply:
There are also subdirectories in the output tree. Not sure but I think we'll know after adding a test for this.

Reply:
If you only want to verify the structure of the output bundle, you can use `archive_contents_test` https://github.com/bazelbuild/rules_apple/blob/22b2c4d1eb45ed41e1c19565a84459e58ce20784/test/starlark_tests/rules/common_verification_tests.bzl#L22

If you want to verify the contents of the output, use `apple_verification_test` 

https://github.com/bazelbuild/rules_apple/blob/22b2c4d1eb45ed41e1c19565a84459e58ce20784/test/starlark_tests/rules/apple_verification_test.bzl#L202

Reply:
Will add some tests this week, thanks for all the help here!

Reply:
Added! It was a little odd to figure out since this outputs a directory not an archive. 

I'm using `AppleBinaryInfo` in the rule for testing so the underlying `archive_contents_test` doesn't attempt to `unzip` the directory.